src/ast.rs

```
@@ -1525,16 +1523,15 @@ impl Parse for CloneSuffix {
         try_begin_parse!("CloneSuffix", ctx, input);
 
         let tail = consume(b".", input)?;
-        let (identifier, tail) = CloneTypeIdentifier::parse(ctx, subs, tail)?;
-        if tail.is_empty() {
-            return Err(error::Error::UnexpectedText);
-        }
```

> On my local machine, and on CI all the tests seem to be passing

That's because I had this one test disabled. I didn't look into it too closely, but the `cppfilt` example does demangle, while the demangle tests did not. They all ran into this `UnexpectedText` error, since the symbol `foo(int) [clone .1988]` only has a nonnegative number but no type identifier.

Coming back to my original message: According to the grammar, the clone type identifier is optional:
```
<clone-suffix> ::= [ . <clone-type-identifier> ] [ . <nonnegative number> ]*
```

With this fix, the number is parsed into the clone type identifier. For demangling, this doesn't really make a difference. However, if you prefer to make the clone type identifier optional, I can do that aswell.

---

Okay I see now. Making it optional seems like the way to go. `c++filt`'s output seems to not care if you're missing either of those, but if you're missing both that's when `c++filt` does nothing.
```
$ c++filt _Z3fooi.12
foo(int) [clone .12]
$ c++filt _Z3fooi.part
foo(int) [clone .part]
$ c++filt _Z3fooi.
_Z3fooi.
```
I think making it optional is the best choice imo. :)

---

I'm trying this now, and I'm no longer sure if it's actually worth the effort. This complicates parsing quite a bit for quite little use. I've pushed this up in a separate commit now to see the difference. I'd highly be in favor of leaving the earlier version, though.

The problem with this implementation is that we have to detect a number of cases throughout the hierarchy of CloneSuffix and CloneTypeIdentifier. This creates very tight coupling, since we have to swallow errors in `zero_or_more`, and then reintroduce redundant checks at the higher level. For example, see https://github.com/gimli-rs/cpp_demangle/pull/194/commits/fbd57414bb6256c0c9f7f08024d39c9691d08709#diff-6a87164025de4254d2818e3ca4bc813aR1538.

`libiberty` Does not seem to differentiate, and instead just consumes everything into one large string. In that sense, the implementation we have right now is probably enough and serves the purpose.

---

Yeah you're right it does add a bit more overhead, and doesn't seem worth it. Okay I think just making it empty should be fine.

---

Thanks. Reverted to the prior implementation.

---

Sorry for making you jump through those hoops, awesome work though!

---

No worries, it's well worth the effort to get a good implementation :)